apache/incubator-metron
# METRON-590 Enable Use of Event Time in Profiler

## [METRON-590](https://issues.apache.org/jira/browse/METRON-590)

### Usage

#### Creating Profiles with Live Data

The most common Profiler use case is creating profiles on live, streaming data. In this case the processing time and event time will normally remain close, but could differ under abnormal conditions.  Event time processing has some significant advantages in this scenario. 

* Profiles are not skewed by high demand that might delay the Profiler's processing.
* Allows the Profiler to take planned/unplanned outages and pick up where it left off
* Produces more accurate behavioral profiles when there is a time difference between when a behavior occurs and when the telemetry produced to tell us about that behavior is received. For example, think of a sensor that collects data in batches or mini-batches where we get data on regular intervals; every 10 minutes, hourly, etc.

#### Creating Profiles with Replayed Data

The other use case that this positions us for is creating profiles from replayed or reprocessed archival data. 

 * For example, I am creating a model based on a new feature that the Profiler is generating for me. When I move that model into Production, I need a historical view of that feature, to train my model. I can replay archived telemetry through the Profiler generating that history of my new feature. This allows the model to be effective in Production on day 1.  

It is important to note that this PR doesn't actually deliver all we need to handle replaying data. This just provides one critical component. I don't want to give anyone the impression that this PR allows us to replay data at this point in time.  [METRON-594](https://issues.apache.org/jira/browse/METRON-594) is tracking the greater effort of replaying telemetry data through the Profiler.

### Changes

* Added event time processing support to the Profiler.  Previously the Profiler only supported processing time aka wall clock time processing.  Event time processing is advantageous as it is not susceptible to skew caused by heavy processing load, allows the reprocessing/replay of archived telemetry data, and under certain circumstances can produce a more accurate profile of entity behavior.

* By default, the Profiler will use event time processing.  The Flux topology definition file must be edited to switch the Profiler to wall clock or processing time.

* The Profiler is now leveraging Storm's windowing functionality introduced in Storm 1.x.  This provides the core engine for event time processing.  This also provides a means for the use of different window types, like sliding windows, in the Profiler.  This is currently not exposed to users of the Profiler as the Flux topology definition file must be edited to use different window types.

* Enhanced the Profiler integration tests which was enabled by the use of event time processing.  The integration tests now generate 24 hours of telemetry data at roughly 3 messages per minute, and then flush profile values every 15 minutes.  The entire stream of values generated by the Profiler is then validated for correctness.

*  Created a `ConfigurationManager` that can be used to read the latest configuration changes in a remote data store like Zookeeper.  The default implementation, `ZkConfigurationManager` replicates the functionality that is embedded in the `ConfiguredBolt` base class.  The Profiler bolts can no longer subclass `ConfiguredBolt` as it subclasses Storm's `BaseRichBolt` which will not work for the Profiler bolts.

* The usability of the Profiler was enhanced to better support active profiles that are subsequently edited by the user. Changes should be handled seamlessly by the Profiler.  This is especially helpful when a mistake is made when creating a profile, which then needs to be fixed and updated.  The Profiler was also made more resilient to failures specific to a single Profile or Tuple.  Individual failures should not impact other Profiles or Tuples.

### Testing

Tested on a multi-node AWS cluster and the Quick Dev environment. Created, edited, and deleted multiple profile definitions as the Profiler was running and responding to the changes.  

Here are manual steps that can be followed to test this PR.  At a high-level, we setup a Metron environment and run it for a period of time to create an 'enrichments' topic containing test data.  We then wait a period of time so that the data becomes stale and the timestamps attached to that data have passed.  We then start the Profiler, but tell it to consume from the `START` of the topic so that it reads all of the stale data.  We then query the Profiler to ensure that it is using the 'old' event timestamp.

* Setup the Environment
    * Start with a clean Metron environment.  
    * Stop all sensors and topologies.
    * Start consuming data from one sensor, like Snort, by starting Pcap Replay, Snort, and the Snort parser topology.  
    * Ensure that data is landing in the 'enrichments' topic.
    * Allow the environment to accumulate data in the 'enrichments' topic for 15 minutes.
    * Stop all topologies and wait for an additional 15 - 60 minutes.
* Setup the Profiler
    * Create the Profiler table in HBase.
        * `create 'profiler', 'P'`
    * Configure the Profiler topology by editing `config/profiler.properties`.
        * `profiler.period.duration=1`  
        * `profiler.input.topic=enrichments`
        * `kafka.start=START`
    * Configure the Profiler Client by adding the following to the global properties.
       * `"profiler.client.period.duration": "1",`
       * `"profiler.client.period.duration.units": "MINUTES"`
    * Define a profile that simply counts messages.
    ```
       {
         "profiles" : [
            {
              "profile": "counter",
              "foreach": "'counter'",
              "update": { "s" : "STATS_ADD(s, 1)" },
              "result": "s"
            }
         ]
       }
    ```
    * Start the Profiler topology.
* Validate
    * Query the Profiler for data with a current timestamp.  There should be none.
    ```
    PROFILE_GET_FROM('counter', 'counter', MILLIS(5, 'MINUTES'))
    ```
    * Query the Profiler for data from 15 - 60 minutes ago, depending on how long you waited.  Data should be returned.
    ```
    PROFILE_GET_FROM('counter', 'counter', MILLIS(5, 'MINUTES'), NOW() - MILLIS(60, 'MINUTES')
    ```
